x-pack/metricbeat/module/statsd/server/server.go

```diff
@@ -163,30 +163,36 @@ func buildMappings(config []StatsdMapping) (map[string]StatsdMapping, error) {
 	return mappings, nil
 }
 
+// It processes metric groups, applies event mappings, and creates Metricbeat events.
+// The generated events include metric fields, labels, and the namespace associated with the MetricSet.
+// Returns a slice of Metricbeat events.
 func (m *MetricSet) getEvents() []*mb.Event {
 	groups := m.processor.GetAll()
-	events := make([]*mb.Event, len(groups))
 
-	for idx, tagGroup := range groups {
-
-		mapstrTags := mapstr.M{}
+	// If there are no groups, return an empty slice.
+	if len(groups) == 0 {
+		return nil
+	}
+	events := make([]*mb.Event, 0, len(groups))
+	for _, tagGroup := range groups {
+		mapstrTags := make(mapstr.M, len(tagGroup.tags))
 		for k, v := range tagGroup.tags {
 			mapstrTags[k] = v
 		}
 
-		sanitizedMetrics := mapstr.M{}
 		for k, v := range tagGroup.metrics {
-			eventMapping(k, v, sanitizedMetrics, m.mappings)
-		}
+			// Apply event mapping to the metric and get MetricSetFields.
+			ms := eventMapping(k, v, m.mappings)
 
-		if len(sanitizedMetrics) == 0 {
-			continue
-		}
-
-		events[idx] = &mb.Event{
-			MetricSetFields: sanitizedMetrics,
-			RootFields:      mapstr.M{"labels": mapstrTags},
-			Namespace:       m.Module().Name(),
+			// If no MetricSetFields were generated, continue to the next metric.
+			if len(ms) == 0 {
+				continue
+			}
+			events = append(events, &mb.Event{
+				MetricSetFields: ms,
+				RootFields:      mapstr.M{"labels": mapstrTags},
```

Comment:
do we need to worry about increasing storage cost by duplicating the tags across many more metric documents?

Reply:
Not all tags are replicated across every metric document. Only the tags relevant to a metric field are included in an event. This is essential for maintaining accurate data relationships.

Furthermore, I've enhanced the code by adding a condition to prevent the inclusion of labels when `mapstrTags` is `nil`. This might lead to reduced storage costs as it avoids storing empty `RootFields` for events with `nil` `mapstrTags`.

Reply:
These tags aren't always there. We're not adding them ourselves; they are optional. In fact, I couldn't find them in the case of Airflow, so there is no extra overhead involved. More about StatsD [tags](https://github.com/statsd/statsd#usage) in general, as they are supported by every StatsD implementation, such as [Datadog](https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/?tab=metrics#) and [Atlassian](https://github.com/atlassian/gostatsd#sending-metrics). Same we are supporting this here.